### PR TITLE
websocket: http request header Connection may contains multiple values

### PR DESCRIPTION
**Pull request by tangxinfa**
http request header `Connection` may contains multiple values, such as: `keep-alive, Upgrade`

*Description*:
Don't use "strcasecmp" to do strict match,  but use "strcasestr" to do partial match, this fix the 400 error response in firefox, because the `Connection` and `Upgrade` request header removed by envoy, become a invalid Websocket request.

### CODE REVIEW

**Comment by mattklein123** [comment]
@tangxinfa thank you for the fix! We will need DCO, format fix, and some tests. Please let us know if you are willing to do those things. cc @rshriram

File: source/common/http/utility.cc [hunk 1]
```diff
@@ -185,9 +185,7 @@ bool Utility::isInternalRequest(const HeaderMap& headers) {
 
 bool Utility::isWebSocketUpgradeRequest(const HeaderMap& headers) {
   return (headers.Connection() && headers.Upgrade() &&
-          (0 == StringUtil::caseInsensitiveCompare(
-                    headers.Connection()->value().c_str(),
-                    Http::Headers::get().ConnectionValues.Upgrade.c_str())) &&
+          (NULL != strcasestr(headers.Connection()->value().c_str(), Http::Headers::get().ConnectionValues.Upgrade.c_str())) &&
```
Comment (htuch) [inline source/common/http/utility.cc]: Don't we still want this to be case insensitive? Maybe we should also do a split operation on `,`, it would be more robust, at the expense of performance.

Comment (ggreenway) [reply source/common/http/utility.cc]: I was thinking the same thing about splitting on comma.  I was looking for a helper/utility that would do this, and I don't see one (which surprised me).  

Perhaps we can add a method to HeaderString for bool entryContains(const LowerCaseString& str) const;

Comment (htuch) [reply source/common/http/utility.cc]: @ggreenway we have https://github.com/envoyproxy/envoy/blob/master/source/common/common/utility.h#L131 and when we gain absl in #1974 we could also use https://github.com/abseil/abseil-cpp/blob/master/absl/strings/str_split.h

Comment (ggreenway) [reply source/common/http/utility.cc]: In this case, do you think the perf cost will be too high to split it and return it?  It would be faster to search in place, but the interface would be less-nice and less-reusable.

Comment (htuch) [reply source/common/http/utility.cc]: I agree that in general we shouldn't copy and split on the hot path when serving reqs. In this particular case, it might be OK to do this perhaps, since we already know we're most likely doing an upgrade due to the earlier test for the Upgrade: header, and if we're doing a Websockets upgrade, it's probably fine to pay a tiny bit extra for what is presumably a long lived session ¯\_(ツ)_/¯

Comment (ggreenway) [reply source/common/http/utility.cc]: I'm fine with that.  If we profile and find we're spending too much time there, we can fix it then.

Comment (mattklein123) [reply source/common/http/utility.cc]: yeah in this case because we are already guarded by the existence of the upgrade header I think it's fine to split after that fact.

Comment (rshriram) [reply source/common/http/utility.cc]: Strcasestr is case insensitive. The question here is whether to abstract this into stringutil or not. This particular case, given that the contributor is doing a one off contribution, I am reluctant to insist on him debugging more :). I can do the follow up and clean it up as it’s not bad as is. But I leave it to you guys to decide on this.

**Comment by rshriram** [comment]
Thank you for testing, tracking down the issue and fixing this. Would you mind copy pasting the explanation in the PR into a comment above the if block, so that others would understand why substring search is being used?

With regard to the tests,  tweak one of the websocket integration tests in test/integration/ and change the connection upgrade to a keepalive,upgrade. This will test if any other piece in Envoy breaks (none should as we use tcp upstream).

File: source/common/http/utility.cc [hunk 2]
```diff
@@ -185,12 +185,10 @@ bool Utility::isInternalRequest(const HeaderMap& headers) {
...
-          (0 == StringUtil::caseInsensitiveCompare(
-                    headers.Connection()->value().c_str(),
-                    Http::Headers::get().ConnectionValues.Upgrade.c_str())) &&
-          (0 == StringUtil::caseInsensitiveCompare(
-                    headers.Upgrade()->value().c_str(),
-                    Http::Headers::get().UpgradeValues.WebSocket.c_str())));
+          headers.Connection()->value().caseInsensitiveContains(
+                  Http::Headers::get().ConnectionValues.Upgrade.c_str()) &&
+          headers.Upgrade()->value().caseInsensitiveContains(
+                  Http::Headers::get().UpgradeValues.WebSocket.c_str()));
```
Comment (rshriram) [inline source/common/http/utility.cc]: Nit: please copy paste the explanation you had for doing the token comparison (about Firefox sending multiple tokens in the connection string)

File: source/common/http/conn_manager_utility.cc [hunk 3]
```diff
@@ -25,7 +25,17 @@ void ConnectionManagerUtility::mutateRequestHeaders(
...
+  // it's means content-length is "0", reference from rfc7230:
+  //   6.  If this is a request message and none of the above are true, then
+  //       the message body length is zero (no message body is present).
+  // Down layers will think it's chunked encoding if no content-length header exists,
+  // we can fix it by explicitly add a "content-length: 0" header here.
+  bool contentLength0 = (!request_headers.TransferEncoding() && !request_headers.ContentLength());
   request_headers.removeTransferEncoding();
+  if (contentLength0) {
+    request_headers.insertContentLength().value(uint64_t(0));
+  }
```
Comment (rshriram) [inline source/common/http/conn_manager_utility.cc]: This seems unrelated to this PR. What is the rationale for this?

File: include/envoy/http/header_map.h [hunk 4]
```diff
@@ -102,6 +102,28 @@ class HeaderString {
...
+  /**
+   * HeaderString is in token list form, each token separated by commas or whitespace,
+   * see https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.1 for more information,
+   * header field value's case sensitivity depends on each header.
+   * @return whether contains token in case insensitive manner.
+   */
+  bool caseInsensitiveContains(const char* token) const {
+    const int n = strlen(token);
+    if (n == 0) {
+      return true;
```
Comment (mattklein123) [inline include/envoy/http/header_map.h]: Do we have a test case for this branch?

File: source/common/http/conn_manager_utility.cc [hunk 5]
```diff
@@ -22,18 +22,31 @@ void ConnectionManagerUtility::mutateRequestHeaders(
...
-  if (protocol != Protocol::Http11 || !Utility::isWebSocketUpgradeRequest(request_headers)) {
+  if (protocol == Protocol::Http11 && Utility::isWebSocketUpgradeRequest(request_headers)) {
+    // Current WebSocket implementation re-use the HTTP1 codec to send upgrade headers to
+    // upstream host, which add "transfer-encoding: chunked" request header if stream not end
+    // and content-length not exists.
+    // In HTTP1.1, if transfer-encoding and content-length both not exists means no request body,
+    // after transfer-encoding striped here, the upstream request become invalid:
+    //   chunked encoding with no request body.
+    // We can fix it by explicitly add a "content-length: 0" request header here.
+    bool no_body = (!request_headers.TransferEncoding() && !request_headers.ContentLength());
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: nit: `const bool`

File: source/common/http/conn_manager_utility.cc [hunk 6]
```diff
@@ -22,18 +22,31 @@ void ConnectionManagerUtility::mutateRequestHeaders(
...
+    if (no_body) {
+      request_headers.insertContentLength().value(uint64_t(0));
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: Do we have a test case where we check for the existence of content-length: 0? We should be able to look for this in one of the conn_manager_impl websocket tests.

File: include/envoy/http/header_map.h [hunk 7]
```diff
@@ -102,6 +102,28 @@ class HeaderString {
...
+    }
+    const char* tokens = c_str();
+    const char* separators = " ,";
+    for (const char* p = tokens; (p = strcasestr(p, token)); p += n) {
+      if ((p == tokens || strchr(separators, *(p - 1)) != NULL) &&
```
Comment (mattklein123) [inline include/envoy/http/header_map.h]: nit: s/NULL/nullptr in both places.

File: include/envoy/http/header_map.h [hunk 8]
```diff
@@ -102,6 +102,28 @@ class HeaderString {
...
+   * @return whether contains token in case insensitive manner.
+   */
+  bool caseInsensitiveContains(const char* token) const {
+    const int n = strlen(token);
+    if (n == 0) {
+      return true;
+    }
+    const char* tokens = c_str();
+    const char* separators = " ,";
+    for (const char* p = tokens; (p = strcasestr(p, token)); p += n) {
```
Comment (mattklein123) [inline include/envoy/http/header_map.h]: this logic is really hard to quickly read. Can you add a small comment on what it is doing?

File: source/common/http/conn_manager_utility.cc [hunk 9]
```diff
@@ -22,18 +22,31 @@ void ConnectionManagerUtility::mutateRequestHeaders(
...
-  request_headers.removeEnvoyInternalRequest();
-  request_headers.removeKeepAlive();
-  request_headers.removeProxyConnection();
-  request_headers.removeTransferEncoding();
 
   // If this is a WebSocket Upgrade request, do not remove the Connection and Upgrade headers,
   // as we forward them verbatim to the upstream hosts.
-  if (protocol != Protocol::Http11 || !Utility::isWebSocketUpgradeRequest(request_headers)) {
+  if (protocol == Protocol::Http11 && Utility::isWebSocketUpgradeRequest(request_headers)) {
+    // Current WebSocket implementation re-use the HTTP1 codec to send upgrade headers to
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: Grammar nits, please just replace with (reflow as needed to 100 col):
```
// The current WebSocket implementation re-uses the HTTP1 codec to send upgrade headers to
// the upstream host. This adds the "transfer-encoding: chunked" request header if the stream 
// has not ended and content-length does not exist. In HTTP1.1, if transfer-encoding and 
// content-length both do not exist this means there is no request body. After transfer-encoding 
// is stripped here, the upstream request becomes invalid. We can fix it by explicitly adding a 
// "content-length: 0" request header here.
```

File: include/envoy/http/header_map.h [hunk 10]
```diff
@@ -102,6 +102,31 @@ class HeaderString {
...
+    const int n = strlen(token);
+    if (n == 0) {
+      return true;
+    }
+
+    // Find token substring, skip if it's partial of other token.
+    const char* tokens = c_str();
+    for (const char* p = tokens; (p = strcasestr(p, token)); p += n) {
+      if ((p == tokens || *(p - 1) == ' ' || *(p - 1) == ',') &&
+          (*(p + n) == '\0' || *(p + n) == ' ' || *(p + n) == ',')) {
```
Comment (mattklein123) [inline include/envoy/http/header_map.h]: Don't you need to check here that the length of the remaining header is >= n before you do this? I think this code can overrun the header. Please make sure you add more test cases to cover small headers.

File: include/envoy/http/header_map.h [hunk 11]
```diff
@@ -102,6 +102,31 @@ class HeaderString {
...
+   * HeaderString is in token list form, each token separated by commas or whitespace,
+   * see https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.1 for more information,
+   * header field value's case sensitivity depends on each header.
+   * @return whether contains token in case insensitive manner.
+   */
+  bool caseInsensitiveContains(const char* token) const {
+    // Avoid dead loop if token argument is empty.
+    const int n = strlen(token);
+    if (n == 0) {
+      return true;
```
Comment (mattklein123) [inline include/envoy/http/header_map.h]: Seems like this should return false?

File: test/integration/integration_test.cc [hunk 12]
```diff
@@ -193,31 +193,35 @@ TEST_P(IntegrationTest, WebSocketConnectionDownstreamDisconnect) {
...
   // The request path gets rewritten from /websocket/test to /websocket.
-  // The size of headers received by the destination is 225 bytes.
+  // The size of headers received by the destination is 228 bytes.
   tcp_client->write(upgrade_req_str);
   fake_upstream_connection = fake_upstreams_[0]->waitForRawConnection();
-  fake_upstream_connection->waitForData(225);
+  const std::string data = fake_upstream_connection->waitForData(228);
+  // The chunked transfer-encoding for no body upstream request must valid
+  if (std::string::npos != data.find("transfer-encoding: chunked")) {
+    EXPECT_NE(std::string::npos, data.find("\r\n\r\n0\r\n"));
```
Comment (mattklein123) [inline test/integration/integration_test.cc]: It's not really clear what this is testing for, can you search for the entire "content-length" etc. header? Same below.

File: include/envoy/http/header_map.h (hunk 11 above)
Comment (tangxinfa) [reply include/envoy/http/header_map.h]: Yes, you are right.

File: include/envoy/http/header_map.h (hunk 10 above)
Comment (tangxinfa) [reply include/envoy/http/header_map.h]: Because it advance the pointer p only if found the token, there is no overrun. I will add more test cases cover small headers.

File: test/integration/integration_test.cc (hunk 12 above)
Comment (tangxinfa) [reply test/integration/integration_test.cc]: Done

File: include/envoy/http/header_map.h (hunk 10 above)
Comment (mattklein123) [reply include/envoy/http/header_map.h]: Ah I see. Sorry got it. Yup you are right. Please just add some small header tests and we are good.

File: source/common/http/conn_manager_utility.cc [hunk 13]
```diff
@@ -22,18 +22,30 @@ void ConnectionManagerUtility::mutateRequestHeaders(
     Runtime::RandomGenerator& random, Runtime::Loader& runtime,
     const LocalInfo::LocalInfo& local_info) {
   // Clean proxy headers.
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: oops sorry, one more nit: Can you please move this comment down with the rest of header removes?
